File: samples/java/DeliveryDispatch/Server/Dispatch/src/main/java/systems/zlink/samples/deliverydispatch/server/dispatch/DispatchHttpServer.java
package systems.zlink.samples.deliverydispatch.server.dispatch;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import systems.zlink.framework.channels.ZLinkClient;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleNames;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.util.Map;

public final class DispatchHttpServer implements AutoCloseable {
    private final HttpServer server;
    private final ObjectMapper json;
    private final ZLinkClient channels;
    private final DispatchWorkQueue queue;

    public DispatchHttpServer(
            ObjectMapper json, ZLinkClient channels, DispatchWorkQueue queue, String endpoint)
            throws IOException {
        this.json = json;
        this.channels = channels;
        this.queue = queue;
        URI uri = URI.create(endpoint);
        this.server = HttpServer.create(new InetSocketAddress(uri.getHost(), uri.getPort()), 0);
        server.createContext("/deliveries", this::handleCreateDelivery);
        server.createContext("/self-check/assert", this::handleServerAssertion);
        server.start();
    }

    // --8<-- [start:doc-dd-http-create]
    private void handleCreateDelivery(HttpExchange exchange) throws IOException {
        if (!"POST".equals(exchange.getRequestMethod())) {
            write(exchange, 405, "");
            return;
        }
        Messages.CreateDeliveryReq request =
                json.readValue(exchange.getRequestBody(), Messages.CreateDeliveryReq.class);
        Messages.AssignDeliveryMsg assign =
                new Messages.AssignDeliveryMsg(
                        request.deliveryId(),
                        request.customerId(),
                        request.pickupAddress(),
                        request.dropoffAddress());
        channels.sendToChannel(SampleNames.DispatchChannel, assign).submit_sync();
        write(
                exchange,
                200,
                json.writeValueAsString(new Messages.CreateDeliveryRes(request.deliveryId())));
    }

    // --8<-- [end:doc-dd-http-create]

    private void handleServerAssertion(HttpExchange exchange) throws IOException {
        if (!"POST".equals(exchange.getRequestMethod())) {
            write(exchange, 405, "");
            return;
        }
        Messages.ServerAssertionReq request =
                json.readValue(exchange.getRequestBody(), Messages.ServerAssertionReq.class);
        queue.assertServerEvidence(request)
                .whenComplete(
                        (response, error) -> {
                            try {
                                if (error != null) {
                                    write(
                                            exchange,
                                            500,
                                            json.writeValueAsString(
                                                    Map.of("error", error.getMessage())));
                                } else {
                                    write(
                                            exchange,
                                            response.passed() ? 200 : 500,
                                            json.writeValueAsString(response));
                                }
                            } catch (IOException writeError) {
                                exchange.close();
                            }
                        });
    }

    private static void write(HttpExchange exchange, int status, String body) throws IOException {
        byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().add("content-type", "application/json");
        exchange.sendResponseHeaders(status, bytes.length);
        exchange.getResponseBody().write(bytes);
        exchange.close();
    }

    @Override
    public void close() {
        server.stop(0);
    }
}

File: samples/kotlin/DeliveryDispatch/Server/Dispatch/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/dispatch/DispatchHttpServer.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.dispatch

import com.fasterxml.jackson.databind.ObjectMapper
import com.sun.net.httpserver.HttpExchange
import com.sun.net.httpserver.HttpServer
import java.net.InetSocketAddress
import java.net.URI
import java.nio.charset.StandardCharsets
import kotlinx.coroutines.future.await
import kotlinx.coroutines.runBlocking
import systems.zlink.framework.channels.ZLinkClient
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleNames
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleTopology
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.AssignDeliveryMsg
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.CreateDeliveryReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.CreateDeliveryRes
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.ServerAssertionReq

class DispatchHttpServer(
    private val json: ObjectMapper,
    private val channels: ZLinkClient,
    private val queue: DispatchWorkQueue,
) : AutoCloseable {
    private val server: HttpServer

    init {
        val endpoint = URI.create(SampleTopology.DispatchHttpEndpoint)
        server = HttpServer.create(InetSocketAddress(endpoint.host, endpoint.port), 0)
        server.createContext("/deliveries", ::handleCreateDelivery)
        server.createContext("/self-check/assert", ::handleServerAssertion)
        server.start()
    }

    // --8<-- [start:doc-dd-http-create]
    private fun handleCreateDelivery(exchange: HttpExchange) {
        if (exchange.requestMethod != "POST") {
            exchange.writeJson(405, "")
            return
        }
        val request = json.readValue(exchange.requestBody, CreateDeliveryReq::class.java)
        val assign =
            AssignDeliveryMsg(
                deliveryId = request.deliveryId,
                customerId = request.customerId,
                pickupAddress = request.pickupAddress,
                dropoffAddress = request.dropoffAddress,
            )
        runBlocking { channels.sendToChannel(SampleNames.DispatchChannel, assign).submit().await() }
        exchange.writeJson(200, json.writeValueAsString(CreateDeliveryRes(request.deliveryId)))
    }

    // --8<-- [end:doc-dd-http-create]

    private fun handleServerAssertion(exchange: HttpExchange) {
        if (exchange.requestMethod != "POST") {
            exchange.writeJson(405, "")
            return
        }
        val request = json.readValue(exchange.requestBody, ServerAssertionReq::class.java)
        val response = queue.assertServerEvidence(request)
        exchange.writeJson(if (response.passed) 200 else 500, json.writeValueAsString(response))
    }

    private fun HttpExchange.writeJson(status: Int, body: String) {
        val bytes = body.toByteArray(StandardCharsets.UTF_8)
        responseHeaders.add("content-type", "application/json")
        sendResponseHeaders(status, bytes.size.toLong())
        responseBody.use { it.write(bytes) }
        close()
    }

    override fun close() {
        server.stop(0)
    }
}
